src/sinks/util/http2.rs

```diff
@@ -26,6 +33,136 @@ pub type Response = http02::Response<Bytes>;
+/// to how `Sink` works. This is because we must "encode" the type
+/// to be able to send it to the inner batch type and sink. Because of
+/// this we must provide a single buffer slot. To ensure the buffer is
+/// fully flushed make sure `poll_complete` returns ready.
+pub struct BatchedHttpSink<T, B, L = HttpRetryLogic>
+where
+    B: Batch,
+    B::Output: Clone + Send + 'static,
+    L: RetryLogic<Response = http02::Response<Bytes>> + Send + 'static,
+{
+    sink: Arc<T>,
+    inner: TowerBatchedSink<HttpBatchService<B::Output>, B, L, B::Output>,
+    // An empty slot is needed to buffer an item where we encoded it but
+    // the inner sink is applying back pressure. This trick is used in the `WithFlatMap`
+    // sink combinator. https://docs.rs/futures/0.1.29/src/futures/sink/with_flat_map.rs.html#20
+    slot: Option<B::Input>,
+}
+
+impl<T, B> BatchedHttpSink<T, B, HttpRetryLogic>
+where
+    B: Batch,
+    B::Output: Clone + Send + 'static,
+    T: HttpSink<Input = B::Input, Output = B::Output>,
+{
+    pub fn new(
+        sink: T,
+        batch: B,
+        request_settings: TowerRequestSettings,
+        batch_settings: BatchSettings,
+        tls_settings: impl Into<MaybeTlsSettings>,
+        cx: &SinkContext,
+    ) -> Self {
+        Self::with_retry_logic(
+            sink,
+            batch,
+            HttpRetryLogic,
+            request_settings,
+            batch_settings,
+            tls_settings,
+            cx,
+        )
+    }
+}
+
+impl<T, B, L> BatchedHttpSink<T, B, L>
+where
+    B: Batch,
+    B::Output: Clone + Send + 'static,
+    L: RetryLogic<Response = http02::Response<Bytes>, Error = hyper13::Error> + Send + 'static,
+    T: HttpSink<Input = B::Input, Output = B::Output>,
+{
+    pub fn with_retry_logic(
+        sink: T,
+        batch: B,
+        logic: L,
+        request_settings: TowerRequestSettings,
+        batch_settings: BatchSettings,
+        tls_settings: impl Into<MaybeTlsSettings>,
+        cx: &SinkContext,
+    ) -> Self {
+        let sink = Arc::new(sink);
+        let sink1 = sink.clone();
+        let svc =
+            HttpBatchService::new(cx.resolver(), tls_settings, move |b| sink1.build_request(b));
+
+        let inner = request_settings.batch_sink(logic, svc, batch, batch_settings, cx.acker());
+
+        Self {
+            sink,
+            inner,
+            slot: None,
+        }
+    }
+}
+
+impl<T, B, L> Sink for BatchedHttpSink<T, B, L>
+where
+    B: Batch,
+    B::Output: Clone + Send + 'static,
+    T: HttpSink<Input = B::Input, Output = B::Output>,
+    L: RetryLogic<Response = http02::Response<Bytes>> + Send + 'static,
+{
+    type SinkItem = crate::Event;
+    type SinkError = crate::Error;
+
+    fn start_send(&mut self, item: Self::SinkItem) -> StartSend<Self::SinkItem, Self::SinkError> {
+        if self.slot.is_some() {
+            self.poll_complete()?;
+            return Ok(AsyncSink::NotReady(item));
```

Comment:
If `poll_complete` succeeds we should be able to make progress here.

Reply:
This is the same implementation we had before, so I'd like to defer to that instead of changing behavior here? What do you think?

Reply:
https://github.com/timberio/vector/blob/master/src/sinks/util/http.rs#L136

Reply:
Yeah if it's just copied that's fine. Probably still worth taking a look later though.

Reply:
https://github.com/timberio/vector/issues/2577